src/kaleidoscope/driver/Bootloader.h

```diff
@@ -0,0 +1,124 @@
+/* -*- mode: c++ -*-
+ * kaleidoscope::driver::bootloader -- Various bootloader drivers
+ * Copyright (C) 2019  Keyboard.io, Inc
+ *
+ * This program is free software: you can redistribute it and/or modify it under
+ * the terms of the GNU General Public License as published by the Free Software
+ * Foundation, version 3.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include <avr/wdt.h>
+
+namespace kaleidoscope {
+namespace driver {
+namespace bootloader {
+
+class Caterina {
+ public:
+  Caterina() {}
+
+  static inline void resetDevice() {
+    // Set the magic bits to get a Caterina-based device
+    // to reboot into the bootloader and stay there, rather
+    // than run move onward
+    //
+    // These values are the same as those defined in
+    // Caterina.c
+
+    uint16_t bootKey = 0x7777;
+    uint16_t *const bootKeyPtr = reinterpret_cast<uint16_t *>(0x0800);
+
+    // Stash the magic key
+    *bootKeyPtr = bootKey;
+
+    // Set a watchdog timer
+    wdt_enable(WDTO_120MS);
+
+    while (1) {} // This infinite loop ensures nothing else
```

Comment:
One style of infinite loop here (personally, I prefer `while (true)`, but whatever…).